From 6342b6393c1d5c7659c9a5b6e72ef43e079f0d6f Mon Sep 17 00:00:00 2001
From: Kai Xing <kxing@codeaurora.org>
Date: Fri, 11 Dec 2020 12:05:08 +0530
Subject: [PATCH 1/4] waylandsink: support fullscreen for waylandsink for
 gstreamer

Add fullscreen property that makes the sink displayed all across the output

Signed-off-by: Kai Xing <kxing@codeaurora.org>
---
 ext/wayland/gstwaylandsink.c | 68 ++++++++++++++++++++++++++++++++++--
 ext/wayland/gstwaylandsink.h |  1 +
 ext/wayland/wlwindow.c       | 17 +++++----
 ext/wayland/wlwindow.h       |  2 +-
 4 files changed, 78 insertions(+), 10 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 0761304..4105744 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -63,7 +63,11 @@ enum
 {
   PROP_0,
   PROP_DISPLAY,
-  PROP_FULLSCREEN
+  PROP_FULLSCREEN,
+  PROP_X,
+  PROP_Y,
+  PROP_DST_WIDTH,
+  PROP_DST_HEIGHT
 };
 
 GST_DEBUG_CATEGORY (gstwayland_debug);
@@ -213,6 +217,25 @@ gst_wayland_sink_class_init (GstWaylandSinkClass * klass)
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
   gst_type_mark_as_plugin_api (GST_TYPE_WAYLAND_VIDEO, 0);
+
+  g_object_class_install_property (gobject_class, PROP_X,
+      g_param_spec_uint ("x", "X Position", "X postion for the content",
+           0, 1920, 0, G_PARAM_READWRITE | G_PARAM_CONSTRUCT));
+
+  g_object_class_install_property (gobject_class, PROP_Y,
+      g_param_spec_uint ("y", "Y Position", "Y postion for the content",
+           0, 1080, 0, G_PARAM_READWRITE | G_PARAM_CONSTRUCT));
+
+  g_object_class_install_property (gobject_class, PROP_DST_WIDTH,
+      g_param_spec_uint ("width", "Destination Width", "Destination Width "
+          "for the content", 0, 1920, 640,
+          G_PARAM_READWRITE | G_PARAM_CONSTRUCT));
+
+  g_object_class_install_property (gobject_class, PROP_DST_HEIGHT,
+      g_param_spec_uint ("height", "Destination Height", "Destination Height "
+          "for the content", 0, 1080, 480,
+          G_PARAM_READWRITE | G_PARAM_CONSTRUCT));
+
 }
 
 static void
@@ -251,6 +274,26 @@ gst_wayland_sink_get_property (GObject * object,
       g_value_set_boolean (value, sink->fullscreen);
       GST_OBJECT_UNLOCK (sink);
       break;
+    case PROP_X:
+      GST_OBJECT_LOCK (sink);
+      g_value_set_uint (value, sink->video_rectangle.x);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_Y:
+      GST_OBJECT_LOCK (sink);
+      g_value_set_uint (value, sink->video_rectangle.y);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_DST_WIDTH:
+      GST_OBJECT_LOCK (sink);
+      g_value_set_uint (value, sink->video_rectangle.w);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_DST_HEIGHT:
+      GST_OBJECT_LOCK (sink);
+      g_value_set_uint (value, sink->video_rectangle.h);
+      GST_OBJECT_UNLOCK (sink);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -274,6 +317,26 @@ gst_wayland_sink_set_property (GObject * object,
       gst_wayland_sink_set_fullscreen (sink, g_value_get_boolean (value));
       GST_OBJECT_UNLOCK (sink);
       break;
+    case PROP_X:
+      GST_OBJECT_LOCK (sink);
+      sink->video_rectangle.x = g_value_get_uint (value);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_Y:
+      GST_OBJECT_LOCK (sink);
+      sink->video_rectangle.y = g_value_get_uint (value);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_DST_WIDTH:
+      GST_OBJECT_LOCK (sink);
+      sink->video_rectangle.w = g_value_get_uint (value);
+      GST_OBJECT_UNLOCK (sink);
+      break;
+    case PROP_DST_HEIGHT:
+      GST_OBJECT_LOCK (sink);
+      sink->video_rectangle.h = g_value_get_uint (value);
+      GST_OBJECT_UNLOCK (sink);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -718,7 +781,8 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
     if (!sink->window) {
       /* if we were not provided a window, create one ourselves */
       sink->window = gst_wl_window_new_toplevel (sink->display,
-          &sink->video_info, sink->fullscreen, &sink->render_lock);
+          &sink->video_info, &sink->video_rectangle,
+          sink->fullscreen, &sink->render_lock);
       g_signal_connect_object (sink->window, "closed",
           G_CALLBACK (on_window_closed), sink, 0);
     }
diff --git a/ext/wayland/gstwaylandsink.h b/ext/wayland/gstwaylandsink.h
index 7aabb6f..faaa14a 100644
--- a/ext/wayland/gstwaylandsink.h
+++ b/ext/wayland/gstwaylandsink.h
@@ -63,6 +63,7 @@ struct _GstWaylandSink
   gboolean fullscreen;
 
   gchar *display_name;
+  GstVideoRectangle video_rectangle;
 
   gboolean redraw_pending;
   GMutex render_lock;
diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
index 66df0fc..5c9b63a 100644
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -254,7 +254,7 @@ gst_wl_window_ensure_fullscreen (GstWlWindow * window, gboolean fullscreen)
 
 GstWlWindow *
 gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
-    gboolean fullscreen, GMutex * render_lock)
+    const GstVideoRectangle * rectangle, gboolean fullscreen, GMutex * render_lock)
 {
   GstWlWindow *window;
 
@@ -283,12 +283,11 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
     xdg_toplevel_add_listener (window->xdg_toplevel,
         &xdg_toplevel_listener, window);
 
-    gst_wl_window_ensure_fullscreen (window, fullscreen);
-
     /* Finally, commit the xdg_surface state as toplevel */
     window->configured = FALSE;
     wl_surface_commit (window->area_surface);
-    wl_display_flush (display->display);
+    wl_display_dispatch(display->display);
+    wl_display_roundtrip(display->display);
 
     g_mutex_lock (&window->configure_mutex);
     timeout = g_get_monotonic_time () + 100 * G_TIME_SPAN_MILLISECOND;
@@ -300,6 +299,9 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
       }
     }
     g_mutex_unlock (&window->configure_mutex);
+
+    gst_wl_window_ensure_fullscreen (window, fullscreen);
+
   } else if (display->wl_shell) {
     /* go toplevel */
     window->wl_shell_surface = wl_shell_get_shell_surface (display->wl_shell,
@@ -322,9 +324,10 @@ gst_wl_window_new_toplevel (GstWlDisplay * display, const GstVideoInfo * info,
     goto error;
   }
 
-  /* render_rectangle is already set via toplevel_configure in
-   * xdg_shell fullscreen mode */
-  if (!(display->xdg_wm_base && fullscreen)) {
+  if ((rectangle->h != 0) && (rectangle->w != 0) && !fullscreen) {
+    gst_wl_window_set_render_rectangle (window, 0, 0, rectangle->w,
+         rectangle->h);
+  } else {
     /* set the initial size to be the same as the reported video size */
     gint width =
         gst_util_uint64_scale_int_round (info->width, info->par_n, info->par_d);
diff --git a/ext/wayland/wlwindow.h b/ext/wayland/wlwindow.h
index 303c336..b9998c3 100644
--- a/ext/wayland/wlwindow.h
+++ b/ext/wayland/wlwindow.h
@@ -83,7 +83,7 @@ GType gst_wl_window_get_type (void);
 void gst_wl_window_ensure_fullscreen (GstWlWindow * window,
         gboolean fullscreen);
 GstWlWindow *gst_wl_window_new_toplevel (GstWlDisplay * display,
-        const GstVideoInfo * info, gboolean fullscreen, GMutex * render_lock);
+        const GstVideoInfo * info, const GstVideoRectangle * rectangle, gboolean fullscreen, GMutex * render_lock);
 GstWlWindow *gst_wl_window_new_in_surface (GstWlDisplay * display,
         struct wl_surface * parent, GMutex * render_lock);
 
-- 
2.25.1

